### PR TITLE
Feature singly linked list descriptions enhancements

### PR DESCRIPTION
**Pull request by TheeMattOliver**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

#### Description

These changes target the Singly Linked List Data Structure challenge.

Specifically, they deal with enhancing the descriptions to improve readability and aid in comprehension, as the instructions and descriptions were unclear. Grammar errors were corrected, and word choice improved.

The "helper" methods for 'size()' and 'head()' were removed as these were redundant and unnecessary, and the tests for each section were updated to reflect this change. Wherever a test called for the size or head of a list, the length and head properties were used respectively.

### CODE REVIEW

**Comment by raisedadead** [comment]
@TheeMattOliver this needs a rebase and squash, to resolve the conflicts. Can you update the pull request?

Reach us in the [Contributors Chat room](https://gitter.im/FreeCodeCamp/Contributors), if you need any assistance.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 1]
```diff
@@ -893,68 +893,119 @@
       "hints": [],
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Please revert this line of change. We do not need to add empty translation blocks.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 2]
```diff
@@ -893,68 +893,119 @@
...
+        "assert((function(){var test = new LinkedList(); test.add('cat'); return test.head.element === 'cat'}()), 'message: Your <code>LinkedList</code> class should assign the first node you add to the <code>head</code> of your list.');",
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); return test.head.next.element === 'dog'}()), 'message: The previous <code>node</code>'s <code>next</code> property in your <code>LinkedList</code> class should have reference to the newest node you created.');",
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); return test.length === 2}()), 'message: The  <code>length</code> of your <code>LinkedList</code> class should equal the number of nodes in the linked list.');"
       ],
       "solutions": [],
       "hints": [],
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Please revert this line of change. We do not need to add empty translation blocks.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 3]
```diff
@@ -988,36 +1031,70 @@
...
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.remove('cat'); return test.head.element === 'dog'}()), 'message: Your <code>remove</code> method should reassign the <code>head</code> to the second node if the element to remove is the head.');",
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.remove('cat'); return test.length === 1}()), 'message: Your <code>Remove</code> method should decrease the <code>length</code> of the linked list by one for every node removed.');",
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog');test.add('kitten'); test.remove('dog'); return test.head.next.element === 'kitten'}()), 'message: Your <code>Remove</code> method should reassign the reference of the previous node of the removed node to the removed node's <code>next</code> reference ');"
       ],
       "solutions": [],
       "hints": [],
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Please revert this line of change. We do not need to add empty translation blocks.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 4]
```diff
@@ -1051,49 +1120,87 @@
...
+        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); return test.length === 3}()), 'message: Your <code>size</code> method should return the length of the linked list');",
         "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); return test.indexOf('kitten') === 2}()), 'message: Your <code>indexOf</code> method should return the index of the given element.');",
         "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); return test.elementAt(1) === 'dog'}()), 'message: Your <code>elementAt</code> method should return at element at a given index.');"
       ],
       "solutions": [],
       "hints": [],
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Please revert this line of change. We do not need to add empty translation blocks.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 5]
```diff
@@ -1167,71 +1278,32 @@
...
-        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.addAt(0,'cat'); return test.head().element === 'cat'}()), 'message: Your <code>addAt</code> method should reassign <code>head</code> to the new node when the given index is 0.');",
-        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.addAt(0,'cat'); return test.size() === 3}()), 'message: Your <code>addAt</code> method should increase the length of the linked list by one for each new node added to the linked list.');",
-        "assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); return (test.addAt(4,'cat') === false); }()), 'message: Your <code>addAt</code> method should return <code>false</code> if a node was unable to be added.');"
-      ],
-      "solutions": [],
-      "hints": [],
-      "type": "waypoint",
-      "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Please revert this line of change. We do not need to add empty translation blocks.

**Review by raisedadead — CHANGES_REQUESTED** [review]
Please remove the additional translation blocks that are empty.

**Comment by TheeMattOliver** [comment]
Done. Thanks.

**Comment by TheeMattOliver** [comment]
@raisedadead Anything else I need to do?

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 6]
```diff
@@ -893,50 +893,50 @@
       "hints": [],
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {}
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Remove white space, refer out contributing guide to configure the linter for your text editor.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 7]
```diff
@@ -865,10 +865,10 @@
...
+        "Another common data structure you'll run into in computer science is the linked list. A linked list is a linear collection of data elements, called nodes, each of which points to the next. Each node in a linked list contains two key pieces of information: the element itself, and a reference to the next node.",
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware of the other people ahead or behind them.",
         "Instructions",
-        "In the code, we've create two nodes, Kitten and Puppy, and we've manually connected the Kitten node to the Puppy node.",
+        "In our code editor, we've created two nodes, Kitten and Puppy, and we've manually connected the Kitten node to the Puppy node.",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Wrap the words like Kitten, Puppy, Cat, Dog in `<code>...</code>` tags for legibility.

Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: You would want to do this at all places in this pull request, where you are referring variable, keywords, etc. in the challenge seed.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 8]
```diff
@@ -865,10 +865,10 @@
...
+        "Another common data structure you'll run into in computer science is the linked list. A linked list is a linear collection of data elements, called nodes, each of which points to the next. Each node in a linked list contains two key pieces of information: the element itself, and a reference to the next node.",
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware of the other people ahead or behind them.",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Does not concern your pull request, but would you mind changing this to `<hr>`

```diff
- "Instructions",
+ "<hr>",
```

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 7 above)
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: A good place to refer the formatting is :
https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json

Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: You many need to add `<dfn></dfn>` to definations as well, just refer the file above

**Review by raisedadead — CHANGES_REQUESTED** [review]
@TheeMattOliver 

Thanks a lot for taking the time to contribute, we really appreciate the efforts.
I have left some comments, basically related to the formatting.
When you have done addressing the changes, please [squash your commits](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231).
Will take a look and merge, thanks once again

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 6 above)
Comment (TheeMattOliver) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Awesome thanks so much Mrugesh @raisedadead! I committed my changes and if there's anything else I need to do just let me know!

**Comment by TheeMattOliver** [comment]
@raisedadead I had a syntax error. Fixed and squashed.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 9]
```diff
@@ -865,11 +865,12 @@
...
+        "Another common data structure you'll run into in computer science is the <dfn>linked list</dfn>. A linked list is a linear collection of data elements, called 'nodes', each of which points to the next. Each <dfn>node</dfn> in a linked list contains two key pieces of information: the <code>element</code> itself, and a reference to the next <code>node</code>.",
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware of the other people ahead or behind them.",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Delete this line.

```diff
- "Instructions",
```

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 10]
```diff
@@ -899,44 +900,45 @@
...
+        "We'll want to have a way to add items to our linked list, so the first method we'll want to create is the <code>add</code> method.", 
+        "If our list is empty, adding an element to our linked list is straightforward enough: we just wrap that element in a <code>Node</code> class, and we assign that node to the <code>head</code> of our linked list." ,
+        "But what if our list already has one or more members? How do we add an element to the list? Recall that each node in a linked list has a <code>next</code> property. To add a node to the list, find the last node in the list, and point that last node's <code>next</code> property at our new node. (Hint: you know you've reached the end of a linked list when a node's <code>next</code> property is <code>null</node>.)",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Delete this line.

```diff
- "Instructions",
```

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 11]
```diff
@@ -948,13 +950,15 @@
...
+        "Whenever we remove a node from a linked list, it's important that we don't accidentally orphan the rest of the list in doing so. Recall that every node's <code>next</code> property points to the node that follows it in the list. If we're removing the middle element, say, we'll want to make sure that we have a connection from that element's previous node's <code>next</code> property to the middle element's <code>next</code> property (which is the next node in the list!)",
+        "This might sound really confusing, so let's return to the conga line example so we have a good conceptual model. Picture yourself in a conga line, and the person directly in front of you leaves the line. The person who just left the line no longer has her hands on anyone in line--and you no longer have your hands on the person that left. You step forward and put your hands on next person you see.",
+        "If the element we wish to remove is the <code>head</code> element, we reassign the <code>head</code> to the second node of the linked list.",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Delete this line.

```diff
- "Instructions",
```

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 12]
```diff
@@ -1014,10 +1018,12 @@
...
+        "Let's add a few more useful methods to our linked list class. Wouldn't it be useful if we could tell if our list was empty or not, as with our <code>Stack</code> and <code>Queue</code> classes?",
+        "We should also be able to find specific elements in our linked list. Traversing through data structures is something you'll want to get a lot of practice with! Let's create an <code>indexOf</code> method that takes an <code>element</code> as an argument, and returns that element's <code>index</code> in the linked list. If the element is not found in the linked list, return <code>-1</code>.",
+        "Let's also implement a method that does the opposite: an <code>elementAt</code> method that takes an <code>index</code> as an argument and returns the <code>element</code> at the given <code>index</code>. If no <code>element</code> is found, return <code>undefined</code>.",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Delete this line.

```diff
- "Instructions",
```

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 13]
```diff
@@ -1093,11 +1101,15 @@
...
+        "Before we move on to another data structure, let's get a last bit of practice with linked lists.",
+        "Let's write a <code>removeAt</code> method that removes the <code>element</code> at a given <code>index</code>. The method should be called <code>removeAt(index)</code>. To remove an <code>element</code> at a certain <code>index</code>, we'll need to keep a running count of each node as we move along the linked list.",
+        "A common technique used to iterate through the elements of a linked list involves a <dfn>'runner'</dfn>, or sentinel, that 'points' at the nodes that your code is comparing. In our case, starting at the <code>head</code> of our list, we start with a <code>currentIndex</code> variable that starts at <code>0</code>. The <code>currentIndex</code> should increment by one for each node we pass.",
+        "Just like our <code>remove(element)</code> method, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.",
         "Instructions",
```
Comment (raisedadead) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Delete this line.

```diff
- "Instructions",
```

**Review by raisedadead — CHANGES_REQUESTED** [review]

**Comment by raisedadead** [comment]
@TheeMattOliver Great work so far, almost there. Some final touches and we are good to go. Sorry for being picky.. 😅 !

when you have made the changes please amend the commit like this:

```sh
git commit --amend --no-edit
git push origin --force
```

This will let you edit the commit, without creating more.
Good luck, and thanks a lot for helping us out.

If you need any assistance hit us in the [Contributors Chat room](https://gitter.im/FreeCodeCamp/Contributors).

**Comment by TheeMattOliver** [comment]
Done! Thanks @raisedadead. Let me know if you'd like me to make further changes.

**Comment by raisedadead** [comment]
@TheeMattOliver looks like you forgot to do a `git add --all`, the changes are not reflecting.

```sh
git add --all
git commit --amend --no-edit
git push origin --force
```

**Comment by TheeMattOliver** [comment]
Thanks @raisedadead I was wondering about that. I followed your last instructions literally. 

Just pushed again.

👍

**Comment by raisedadead** [comment]
@TheeMattOliver thanks for making the changes. Will merge after a QA check first thing in the morning.

**Comment by TheeMattOliver** [comment]
Hey guys did this get hung up on a check I didn't pass again? @raisedadead

**Comment by raisedadead** [comment]
@TheeMattOliver looks like this got into a merge conflict. Do not worry we will resolve and merge upstream.

**Comment by raisedadead** [comment]
@TheeMattOliver I have resolved the conflicts, but this will break the tests. Mind having a look?

**Comment by raisedadead** [comment]
On you local setup, you need to delete this branch (local machine only, NOT on GitHub) and pull down again.

```
git checkout staging
git branch -D feature-singly-linked-list-descriptions-enhancements
git fetch origin --all
git checkout feature-singly-linked-list-descriptions-enhancements
Make Necessary changes to the tests.
git add --all
git commit --amend --no-edit
git push origin feature-singly-linked-list-descriptions-enhancements --force
```

**Comment by TheeMattOliver** [comment]
@raisedadead Will do now, pls hold.

**Comment by raisedadead** [comment]
Hey @TheeMattOliver ! I guess you would have to make a fresh PR. I couldn't get this working. Sincerest apologies for the inconvenience.

Let me know if you need assistance. 

Thanks a lot once again for your contribution.

**Pull request closed without merging**